### ISSUE TITLE
Lowered letters in Hungarian text

### ISSUE THREAD

**Issue by pnorman** [versatiles-org/versatiles-fonts#7]
I'm seeing some letters in Hungarian text being lowered in the text line
![image](https://github.com/user-attachments/assets/19920d6f-ec24-43f2-9a75-9bf5124c5dbf)

You can see this at https://pnorman.github.io/tilekiln-shortbread-demo/#11.47/44.4357/26.1556 which is using the style provided in the releases

**Comment by hyperknot** [comment]
Hungarian speaker here, those are Romanian letters / cities. Still valid point though.

**Comment by Robot8A** [comment]
Yes, it seems to be happening with Romanian letters [Țț](https://en.wikipedia.org/wiki/%C8%9A) and [Șș](https://en.wikipedia.org/wiki/%C8%98), see attached image
![image](https://github.com/user-attachments/assets/fee319f2-4f15-4f7b-bde7-0af589f1b45b)

**Comment by MichaelKreil** [comment]
I'm not sure what kind of problem this is. My best guess is that it is not a style problem, but a font generation problem, so we might want to move the issue to the [versatiles-fonts repo](https://github.com/versatiles-org/versatiles-fonts).

Currently, we generate the glyphs using the outdated [node-fontnik](https://github.com/mapbox/node-fontnik), which is basically NodeJS wrapped C++ code.
We may want to migrate to the [Rust PBF Font Tools](https://github.com/stadiamaps/sdf_font_tools/tree/main/pbf_font_tools).

But first, does anyone have any ideas on how to debug this problem? Are there any tools for debugging glyphs in PBF format?

**Comment by hyperknot** [comment]
I'm quite sure it's the font, nothing to do with the style. I'm also in the process of making a correct Open Sans + Noto Sans font in MapLibre.
https://github.com/maplibre/maplibre-gl-js/issues/4820

**Comment by hyperknot** [comment]
I've posted my journey so far on the rabbit hole of open source maplibre fonts here:
https://github.com/maplibre/maplibre-gl-js/issues/4820#issuecomment-2478722628

**Comment by MichaelKreil** [comment]
After some research and thought, I finally snapped and built our own glyph rendering engine: https://github.com/versatiles-org/versatiles-glyphs-rs

Here is a preview of the results: https://versatiles.org/versatiles-glyphs-rs/

**Comment by hyperknot** [comment]
Great work @MichaelKreil! One thing not clear to me, how does it differ from https://github.com/maplibre/font-maker?
@bdon said that they put lot of work in optimising the visual output of that.

**Comment by MichaelKreil** [comment]
"Font Maker" uses an unmaintained side branch of MapBox's [sdf-glyph-foundry](https://github.com/mapbox/sdf-glyph-foundry/blob/point-in-polygon/include/mapbox/glyph_foundry_impl.hpp), wrapped in a [C++ main file by MapLibre](https://github.com/maplibre/font-maker/blob/main/main.cpp), wrapped as a [web assembly](https://github.com/maplibre/font-maker/blob/main/build_wasm.sh), wrapped in a [TypeScript project](https://github.com/maplibre/font-maker/). I think this is a good point to start from scratch. :)
I basically learnt from all these repos, but combined them into a single Rust project, fixed the letter offset and added some features. Basically, you can now just throw a folder of font files at it and it will generate a folder with all the glyphs and indexes.

**Comment by hyperknot** [comment]
That's great to hear!

How is the visual quality compared to font-maker? @bdon says that Inter's lower-case “e” is a good test for that.

Details here: https://mastodon.social/@bdon/111107453994158512

**Comment by MichaelKreil** [comment]
[versatiles-glyphs](https://github.com/versatiles-org/versatiles-glyphs-rs) uses the same algorithms as Font Maker. It renders the vector geometry directly, without an intermediate rasterisation step (unlike TinySDF), and processes Bézier curves - just like Font Maker.

**Comment by hyperknot** [comment]
Great, super nice!
